CMakeLists.txt

```diff
@@ -0,0 +1,539 @@
+        crypto/engine/eng_fat.c crypto/bn/rsaz_exp.c crypto/ec/ecp_nistz256.c
+        crypto/sha/sha256-x86_64.s crypto/sha/sha1-mb-x86_64.s crypto/engine/tb_asnmth.c
+        crypto/ec/ecp_nistz256-x86_64.s crypto/engine/eng_ctrl.c crypto/engine/tb_eckey.c
+        crypto/engine/eng_table.c crypto/engine/eng_rdrand.c crypto/bn/x86_64-mont.s
+        crypto/engine/eng_cnf.c crypto/bf/bf_enc.c crypto/sha/sha256-mb-x86_64.s
+        crypto/aes/aesni-sha1-x86_64.s crypto/engine/tb_rand.c crypto/engine/eng_dyn.c
+        crypto/chacha/chacha-x86_64.s crypto/engine/eng_list.c crypto/sha/sha1-x86_64.s
+        crypto/engine/eng_lib.c crypto/engine/tb_cipher.c crypto/aes/aesni-mb-x86_64.s
+        crypto/des/fcrypt_b.c crypto/engine/tb_pkmeth.c crypto/rc4/rc4-md5-x86_64.s
+        crypto/rc4/rc4-x86_64.s crypto/des/des_enc.c crypto/modes/ghash-x86_64.s
+        crypto/sha/keccak1600-x86_64.s crypto/md5/md5-x86_64.s crypto/bn/x86_64-mont5.s
+        crypto/bn/rsaz-avx2.s crypto/engine/eng_all.c crypto/camellia/cmll_misc.c
+        crypto/bn/rsaz-x86_64.s crypto/aes/aes_core.c crypto/engine/tb_digest.c
+        crypto/aes/aesni-sha256-x86_64.s crypto/engine/tb_dsa.c crypto/bn/x86_64-gf2m.s
+        crypto/poly1305/poly1305-x86_64.s crypto/aes/vpaes-x86_64.s crypto/engine/tb_dh.c
+        crypto/engine/eng_pkey.c crypto/engine/eng_openssl.c crypto/sha/sha512-x86_64.s
+        crypto/engine/tb_rsa.c
+    )
+
+    add_library(crypto STATIC ${crypto_SRCS})
+
+    target_include_directories(crypto PRIVATE crypto/modes include ${CMAKE_CURRENT_SOURCE_DIR} crypto/ec/curve448/arch_32 crypto crypto/include crypto/ec/curve448)
+    target_compile_definitions(crypto PRIVATE  MD5_ASM POLY1305_ASM X25519_ASM RC4_ASM OPENSSL_CPUID_OBJ OPENSSLDIR="/usr/local/ssl" ECP_NISTZ256_ASM VPAES_ASM SHA256_ASM OPENSSL_BN_ASM_MONT5 _REENTRANT OPENSSL_IA32_SSE2 SHA1_ASM OPENSSL_PIC SHA512_ASM L_ENDIAN GHASH_ASM ENGINESDIR="/usr/local/lib/engines-1.1" KECCAK1600_ASM OPENSSL_BN_ASM_MONT OPENSSL_BN_ASM_GF2m)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-common -fPIC ")
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fno-common -fPIC ")
+    set_clang_unused()
+
+elseif(CMAKE_OSX_SYSROOT MATCHES iPhoneOS)
+
+    set(arch "ios-cross")
+    message("Building ${arch}")
+
+    find_program(PERL_EXECUTABLE perl)
+    if(NOT PERL_EXECUTABLE)
+        message(FATAL_ERROR("Perl not found"))
+    endif(NOT PERL_EXECUTABLE)
+
+    generate_configdata(${arch})
+    generate_mkbuildinf("clang -arch arm64 -mios-version-min=7.0.0 -fno-common -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer//SDKs/iPhoneOS.sdk -DOPENSSL_PIC -D_REENTRANT" "ios64-cross")
+    generate_headers(${arch})
+    
+    add_library(ssl STATIC ${COMMON_ssl_SRCS})
+
+    target_include_directories(ssl PRIVATE ${CMAKE_CURRENT_SOURCE_DIR} include)
+    target_compile_definitions(ssl PRIVATE  _REENTRANT OPENSSLDIR="/usr/local/ssl" ENGINESDIR="/usr/local/lib/engines-1.1" OPENSSL_PIC)
```

Comment:
since we need a portable build that works the same regardless of whether openssl is installed on the system, `OPENSSLDIR` and `ENGINESDIR` should not be defined

Reply:
I'm not sure it's feasible to do so at this moment (given that everything compiles for us on CI):
1) Performance most likely will suffer (`no-engine` flag essentially means no hardware support). 
2) Everything must be re-generated and in the case if there are any errors, they must be dealt with.

Reply:
1. what provides these engines? I don't see this in build scripts and I'm not sure how Android could do it on its own
2. I don't think that full regeneration is necessary for this, the defines can just be removed to see if everything still builds and runs

Reply:
By "what provides" do you mean how did they appear in CMakeLists in the first place? Perl script.
On the other hand, I just checked and it seems that both of these are empty... at least in WSL bash. I'll combine these changes w/ the common source and let the builds run on Azure to check then.

Reply:
also, I see there's `OPENSSL_NO_STATIC_ENGINE` defined - if anything, I don't see issues with using static engines but should not use dynamic ones

Reply:
> By "what provides" do you mean how did they appear in CMakeLists in the first place?

By that I mean, how does a device end up with having `.so` files in `/usr/local/lib/engines-1.1` that this library could use?

Reply:
Not sure, to be honest, since I don't think Android even has `/usr/` to begin with...

Reply:
It appears, that OPENSSL is needed for **default** location of `/certs` and `ct_log_list.cnf` (which contains Certificate Transparency logs). These locations can be overridden by specifying appropriate environment variables. Anyways, if we're to remove the OPENSSLDIR define completely, then we need to patch the `cryptlib.h`. Not much, like 2-3 lines, but I'd rather we just leave `OPENSSLDIR` as is or at least set it to some safe value.
BTW, it's the same with `ENGINESDIR`, there's one place that depends on this define (for **default** engines location), but it can be overridden by providing an environment variable (see screenshot).

>also, I see there's OPENSSL_NO_STATIC_ENGINE defined - if anything, I don't see issues with using static engines but should not use dynamic ones

Not arguing, but curious: why is dynamic a problem for us specifically?

![image](https://user-images.githubusercontent.com/1312792/72814812-5d715900-3c6e-11ea-958f-bca3c5852383.png)

Reply:
> why is dynamic a problem for us specifically?

this is just the usual portable binary stuff - they should work the same on all computers so that...

1. we know that however it works locally (wrt. performance and capabilities) on the machines that built it, it will work the same exact way on all other computers (or at least the same hardware category, e.g. with or without some instruction set support)
2. we don't get bug reports about exotic situations where things break down - those are the most difficult to debug because it's hard to reproduce them

> Anyways, if we're to remove the OPENSSLDIR define completely, then we need to patch the cryptlib.h.

this is unfortunate, `cversion.c` gave me the impression that not defining them is supported

ideally since we're shipping a self-contained build, we would remove both environment variable checks and defines...

but right now I think it would be enough to

1. disable the code that enumerates the available plugins if it's active (can check with strace or similar tools)
2. set the defines to some value that is extremely difficult to obtain (like non-ascii utf8 folder names in root)

P.S. as for certs, I think they it's safe to let the users specify those in addition to the built-in ones so that the games still have a chance to work after 10 years without updates

Reply:
> disable the code that enumerates the available plugins if it's active (can check with strace or similar tools)

By "plugins" do you mean "engines" and is it a reference to the snippet in the screenshot?

>set the defines to some value that is extremely difficult to obtain (like non-ascii utf8 folder names in root)

So that it wouldn't collide with some directory on a user system by accident?

>I think they it's safe to let the users specify those in addition to the built-in ones

But built-in ones wouldn't be accessible if we do the above-mentioned?

Reply:
> By "plugins" do you mean "engines"

yes

> and is it a reference to the snippet in the screenshot?

not sure if that's the code in question, I just expect there to be code somewhere that iterates the directory and tries to load the .so files in it
after some searching from the other end, it looks like that is indeed the code that ultimately controls the process

> So that it wouldn't collide with some directory on a user system by accident?

yes

> But built-in ones wouldn't be accessible if we do the above-mentioned?

by built-in ones I mean those we have baked in our library (`OWN_CA_CERTS` define), they should be accessible regardless of what we do here

I'm currently not sure how the certificate loading code works in openssl exactly, just saying it seems reasonable to load both the certs from our library and the user's OS certificate stores and use all of them to verify the server

Reply:
After doing a bit more investigating, given that only cURL uses OpenSSL, it seems that `ENGINE_` API is called only if there's `USE_OPENSSL_ENGINE` defined (we do not of course). 
I also looked at example through `strace` and it doesn't show any mention of engine.
Do you think we still need to butcher the `eng_list.c` or shall we leave to future times?

P.S. On a related note, after setting `OPENSSLDIR` to a "highly-unlikely-to-be-existing" directory, the following surfaces in logs:
`openat(AT_FDCWD, "/t\304\223m\321\200/123/openssl.cnf", O_RDONLY) = -1 ENOENT (No such file or directory)`

However, after settings `OPENSSL_CONF` environment variable, a lookup directory is now changed:
`openat(AT_FDCWD, "/usr/ssl/conf.cnf", O_RDONLY) = -1 ENOENT (No such file or directory)`

So it seems that we got ourselves covered on that one.

Reply:
well, if there's no evidence that the dynamic engines are in use, then this can wait for sure
thanks for the research!